source/tests/consistent/common.py

```diff
@@ -0,0 +1,404 @@
+        if self.skip_dp:
+            self.skipTest("Unsupported backend")
+        ref_backend = self.get_reference_backend()
+        if ref_backend == self.RefBackend.DP:
+            self.skipTest("Reference is self")
+        ret1, data1 = self.get_reference_ret_serialization(ref_backend)
+        ret1 = self.extract_ret(ret1, ref_backend)
+        dp_obj = self.dp_class.deserialize(data1)
+        ret2 = self.eval_dp(dp_obj)
+        ret2 = self.extract_ret(ret2, self.RefBackend.DP)
+        data2 = dp_obj.serialize()
+        np.testing.assert_equal(data1, data2)
+        for rr1, rr2 in zip(ret1, ret2):
+            np.testing.assert_allclose(rr1, rr2)
+
+    def test_dp_self_consistent(self):
+        """Test whether DP is self consistent."""
+        if self.skip_dp:
+            self.skipTest("Unsupported backend")
+        obj1 = self.init_backend_cls(self.dp_class)
+        ret1, data1 = self.get_dp_ret_serialization_from_cls(obj1)
+        obj1 = self.dp_class.deserialize(data1)
+        ret2, data2 = self.get_dp_ret_serialization_from_cls(obj1)
+        np.testing.assert_equal(data1, data2)
+        for rr1, rr2 in zip(ret1, ret2):
+            if isinstance(rr1, np.ndarray) and isinstance(rr2, np.ndarray):
+                np.testing.assert_allclose(rr1, rr2)
+            else:
+                self.assertEqual(rr1, rr2)
+
+    def test_pt_consistent_with_ref(self):
+        """Test whether PT and reference are consistent."""
+        if self.skip_pt:
+            self.skipTest("Unsupported backend")
+        ref_backend = self.get_reference_backend()
+        if ref_backend == self.RefBackend.PT:
+            self.skipTest("Reference is self")
+        ret1, data1 = self.get_reference_ret_serialization(ref_backend)
+        ret1 = self.extract_ret(ret1, ref_backend)
+        obj = self.pt_class.deserialize(data1)
+        ret2 = self.eval_pt(obj)
+        ret2 = self.extract_ret(ret2, self.RefBackend.PT)
+        data2 = obj.serialize()
+        np.testing.assert_equal(data1, data2)
+        for rr1, rr2 in zip(ret1, ret2):
+            np.testing.assert_allclose(rr1, rr2)
+
+    def test_pt_self_consistent(self):
+        """Test whether PT is self consistent."""
+        if self.skip_pt:
+            self.skipTest("Unsupported backend")
+        obj1 = self.init_backend_cls(self.pt_class)
+        ret1, data1 = self.get_pt_ret_serialization_from_cls(obj1)
+        obj2 = self.pt_class.deserialize(data1)
+        ret2, data2 = self.get_pt_ret_serialization_from_cls(obj2)
+        np.testing.assert_equal(data1, data2)
+        for rr1, rr2 in zip(ret1, ret2):
+            if isinstance(rr1, np.ndarray) and isinstance(rr2, np.ndarray):
+                np.testing.assert_allclose(rr1, rr2)
+            else:
+                self.assertEqual(rr1, rr2)
+
+    def tearDown(self) -> None:
+        """Clear the TF session."""
+        if not self.skip_tf:
+            clear_session()
+
+
+class DescriptorTest:
```

Comment:
provide in a separate file for descriptors? or shall we have a folder structure for the tests of different components?

Reply:
A folder structure is excellent.